### ISSUE TITLE
KLEE termination on error paths for a property prevents verification of other properties

### ISSUE THREAD

**Issue by delcypher** [delcypher/klee-runner#6]
This is a slight subtlety that occurred to me recently. I don't think this is really an issue but it is something to aware of.

This occurs because KLEE does not consider verification tasks independently from each other. Consequently if it if KLEE finds a bug say for `no_invalid_deref` then it cannot conclude if the program is correct with respect to other tasks (e.g. `no_overshift`) because execution which could continue past the bug is not performed.

This is handled in the code at

https://github.com/delcypher/klee-runner/blob/fe832fa3f78b68cc74a4ecf268cf4ff7fc5a4836/kleeanalysis/analyse.py#L131

Currently if KLEE finds a bug for one property the `kleeanalysis` code will report that KLEE reports that the correctness of all other verification tasks is unknown.

What I've done is very conservative and we could improve this slightly by considering terminating and non-terminating bugs.

For example hitting an assert or abort definitely terminates the program so there can be no other bugs to find past the assert/abort as they are unreachable. Taking this into account would allow us to conclude correct for all other verification tasks if KLEE found only found bugs that are assertion failures or aborts being hit.

Performing things like overshift and an invalid deref are not terminating bugs (program execution might continue even though it is undefined) so if KLEE finds a bug here we can't conclude that other verification tasks are correct.

@ccadar @afd @danielschemmel What do you think? I'm not sure if I've done this right. It is debatable whether we should even worry about bugs that can occur after undefined behaviour has occurred because the rest of the execution is undefined.

Here's a real example. This shows the output of `result-info-summary.py` running on Imperial's tool on Aachen's `aachen/syn/halve_double-noflow.x86_64` benchmark. Note the script concludes that KLEE reports a failure for the `task no_assert_fail` verification task but reports unknown for all other tasks.

```
# of raw results: 1
# of KleeRunnerResult.BAD_EXIT: 0
# of KleeRunnerResult.INVALID_KLEE_DIR: 0
# of KleeRunnerResult.OUT_OF_MEMORY: 0
# of KleeRunnerResult.OUT_OF_TIME: 0
# of KleeRunnerResult.SENTINEL: 0
# of KleeRunnerResult.VALID_KLEE_DIR: 1

=== Verification counts per benchmark ===
# of KleeResultCorrect: 0
# of KleeResultIncorrect: 1
# of KleeResultUnknown: 0

=== Verification counts by task ===
# of KleeResultCorrect: 0
# of KleeResultIncorrect: 1
  # of task no_assert_fail: 1
# of KleeResultUnknown: 5
  # of task no_integer_division_by_zero: 1
  # of task no_invalid_deref: 1
  # of task no_invalid_free: 1
  # of task no_overshift: 1
  # of task no_reach_error_function: 1
  Reasons for reporting unknown
    # because "Cannot verify because KLEE terminated with other counter examples that block further checking of the task": 5

# of total tasks: 6 (1 * 6)
                                                                                                                                                                                                                   === Spec matches by task ===                                                                                                                                                                                       # of KleeResultMatchSpec: 1                                                                                                                                                                                          # of correct: 0                                                                                                                                                                                                    # of incorrect: 1                                                                                                                                                                                                # of KleeResultMismatchSpec: 0                                                                                                                                                                                     # of KleeResultUnknownMatchSpec: 5                                                                                                                                                                                   By reason:
   # because "KLEE could not determine correctness": 5                                                                                                                                                               By expected correctness:
    # of expect correct: 5                                                                                                                                                                                             # of expect incorrect: 0                                                                                                                                                                                           # of expect unknown: 0 
```

**Comment by delcypher** [comment]
This has been partially fixed by e79d0df6416362906a5707c6900e45e2eac5be2c

**Comment by delcypher** [comment]
I think this fixed so closing.
